crates/orchestrator/src/lib.rs

```diff
@@ -269,41 +278,64 @@ where
     #[allow(clippy::cast_possible_truncation)]
     fn register_public_key(
         &mut self,
-        node_index: u64,
         pubkey: &mut Vec<u8>,
         is_da: bool,
-    ) -> Result<(), ServerError> {
-        if self.pub_posted.contains(&node_index) {
+        libp2p_address: Option<Multiaddr>,
+        libp2p_public_key: Option<PeerId>,
+    ) -> Result<u64, ServerError> {
+        if !self.accepting_new_keys {
             return Err(ServerError {
-                status: tide_disco::StatusCode::BadRequest,
-                message: "Node has already posted public key".to_string(),
+                status: tide_disco::StatusCode::Forbidden,
+                message:
+                    "Network has been started manually, and is no longer registering new keys."
+                        .to_string(),
             });
         }
-        self.pub_posted.insert(node_index);
 
-        // The guess is the first extra 12 bytes are from orchestrator serialization
-        pubkey.drain(..12);
-        let register_pub_key_with_stake = PeerConfig::<KEY>::from_bytes(pubkey).unwrap();
-        self.config.config.known_nodes_with_stake[node_index as usize] =
-            register_pub_key_with_stake.clone();
+        if let Some(node_index) = self.pub_posted.get(pubkey) {
+            return Ok(*node_index);
+        }
+
+        let node_index = self.pub_posted.len() as u64;
+
+        self.pub_posted.insert(pubkey.clone(), node_index);
+
+        let staked_pubkey = PeerConfig::<KEY>::from_bytes(pubkey).unwrap();
+        self.config
+            .config
+            .known_nodes_with_stake
+            .push(staked_pubkey.clone());
 
         // If the node wants to be DA, add it to the list of known DAs
         if is_da {
-            self.config
-                .config
-                .known_da_nodes
-                .push(register_pub_key_with_stake);
+            self.config.config.known_da_nodes.push(staked_pubkey);
```

Comment:
As mentioned above, here I guess `is_da` is always `false`, because we set it at a later step.